app/react/V2/Routes/Settings/Thesauri/specs/Thesauri.spec.tsx

```diff
@@ -0,0 +1,469 @@
+        await act(async () => {
+          fireEvent.click(screen.getByText('Add group').parentNode!);
+        });
+        let newItemForm: HTMLElement;
+        await act(async () => {
+          [, newItemForm] = renderResult.container.getElementsByTagName('form');
+          fireEvent.click(screen.getByTestId('thesaurus-form-submit'));
+        });
+        await act(async () => {
+          expect(screen.getAllByText('This field is required')).toHaveLength(2);
+          fireEvent.click(within(newItemForm).getByTestId('thesaurus-form-cancel'));
+        });
+      });
+
+      it('should add items into an existing group', async () => {
+        await act(async () => {
+          fireEvent.click(screen.getByText('Add item').parentNode!);
+        });
+        await act(async () => {
+          const newItemForm = renderResult.container.getElementsByTagName('form')[1];
+          fireEvent.change(within(newItemForm).getByLabelText('Title') as HTMLInputElement, {
+            target: { value: 'new child 2' },
+          });
+          fireEvent.click(within(newItemForm).getAllByLabelText('Group')[0]);
+        });
+        await act(async () => {
+          fireEvent.change(screen.getByLabelText('Group'), {
+            target: {
+              value: (screen.getAllByRole('option')[1] as HTMLInputElement).value,
+            },
+          });
+          fireEvent.click(screen.getByTestId('thesaurus-form-submit'));
+        });
+        await act(async () => {
+          rows = await waitFor(() => screen.getAllByRole('row'));
+          expect(rows.length).toBe(4);
+        });
+      });
+
+      it('should prevent leaving without saving', async () => {
+        fireEvent.click(screen.getByText('Cancel').parentNode!);
+        await act(async () => {
+          expect(screen.getByText('Discard changes')).toBeInTheDocument();
+          fireEvent.click(within(screen.getByTestId('modal')).getByText('Cancel').parentNode!);
+        });
+      });
+
+      it('should save the new thesaurus', async () => {
+        await act(async () => {
+          fireEvent.click(screen.getByText('Save').parentNode!);
+        });
+        await act(async () => {
+          expect(saveFn).toHaveBeenCalledWith(
+            {
+              name: 'new thesaurus',
+              values: [
+                {
+                  label: 'single value 1',
+                },
+                {
+                  label: 'single value 2',
+                },
+                {
+                  label: 'Group 1',
+                  values: [
+                    {
+                      label: 'Child 1',
+                    },
+                    {
+                      label: 'new child 2',
+                    },
+                  ],
+                },
+              ],
+            },
+            undefined
+          );
+        });
+      });
+    });
+
+    describe('Edit Thesaurus', () => {
+      it('should edit a root item', async () => {
```

Comment:
There are some tests that don't have any `expect`s. Is that correct? How do you check if the test fails?

Reply:
It's a good one, I'm using screen.get... functions which fail in case the looken element doesn't exists in the document (a couple use screen.query... which doesn't fail if the element doesn't exist), also @testing-library/react/pure to keep the state between the tests, but adding expects definitively helps a lot with the readability of the tests, I added a function to check the right call of the save function where it applies.